### PR TITLE
Simplify report replay checking.

### LINKED ISSUE #557: Explicit backoff logic for aggregation jobs
As mentioned in #556, replay protection implies a relatively stringent operational requirement for the Helper: many aggregation jobs might make concurrent transactions on the same database, which can easily overwhelm the database if not sufficiently provisioned.

To mitigate this problem, the Helper can cancel an aggregation job and ask the Leader to try again later. One option is to respond with HTTP status 429 and a "retry-after" header indicating when it should be safe to retry.

I'd like to suggest that we spell this out explicitly in the draft.

### PR DESCRIPTION
**Pull request by branlwyd**
Closes #557.

Specifically, we relax the check from the specific report ID, to intersection with intervals of report IDs. This will allow implementations to drop all long-term per-report storage, considerably reducing the storage requirements to implement the protocol. The tradeoff is in increased aggregation latency: typically, aggregation cannot begin until the Leader is sure that no further reports will arrive for a given time window.

### CODE REVIEW

**Comment by branlwyd** [comment]
A few thoughts:
* This specifies proposal (2a) from [Bearing the cost of aggregation](https://docs.google.com/document/d/1LNzDEzY5CUlOlOjzk5Z0fkECDF-5gukV-SVwC1gj6Wo/edit#bookmark=id.c0uo88am22oz), extended to treat the leader-selected batch mode uniformly with the time-interval batch mode (as described in [a comment](https://docs.google.com/document/d/1LNzDEzY5CUlOlOjzk5Z0fkECDF-5gukV-SVwC1gj6Wo/edit?disco=AAABUyvf-TY)).
* I am not so happy with the name "report ID range" for the new data structure described in the PR, as the structure is actually a `(timestamp, report ID interval)`. Thoughts on naming welcome.
* I did a small amount of editorial work in the sections I was already touching; please review closely.

File: draft-ietf-ppm-dap.md
Comment (simon-friedberger) [inline draft-ietf-ppm-dap.md:1292]: Maybe be explicit about the requirement instead "aggregation can begin as soon as the leader is willing to discard any further incoming reports for this time window."

File: draft-ietf-ppm-dap.md
Comment (simon-friedberger) [inline draft-ietf-ppm-dap.md:1316]: I agree that calling the new thing a "report ID range" makes everything hard to read. This sentence looks like it is comparing report ID ranges but it is actually comparing "report ID ranges".

Considering the fact that you're saying "a set of" here. Does that mean the old behavior would still be allowed by making each range have exactly one report ID?

Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1316]: To clarify, I am wondering why anybody would want to do this given that the report IDs are randomly assigned. Is the intent to allow people to recover from some amount of late arriving reports?

Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1316]: How strict do you think we should be? As we discussed on slack today, at a minimum we want the helper to be able to reject any range that overlaps with an already aggregated range. But the helper could make a smarter decision by "merging" the two ranges and rejecting only the overlap, in which case the MUST here would not be enforced by the helper.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1733]: Let's make sure this is harmonized with the rest of the draft. I believe we say "truncate", not "round". However I haven't checked.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:2433]: Note that we never use "&" in place of "and" in this draft.
```suggestion
## Aggregation Jobs and Replay Protection
```

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:2437]: nit: consistency. I'd be cool with saying "protocol" instead of "sub-protocol", but let's track in an issue.
```suggestion
As part of the aggregation sub-protocol, the Leader is responsible for generating
```

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:2439]: nit: consistency. FWIW I weakly prefer "overlap" to "intersect".
```suggestion
non-intersecting intervals of report IDs for each time window (see {{agg-init}}).
```

Comment (cjpatton) [inline draft-ietf-ppm-dap.md:2439]: Let's make sure "time window" is defined in the terms section of the doc

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:2513]: I believe the old text still applies: a batch interval is a sequence of time windows; and the batch interval is not known at aggregation time.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:2477]: I think the core idea here still applies: it's just that we're storing report ranges rather than IDs.

File: draft-ietf-ppm-dap.md
Comment (branlwyd) [reply draft-ietf-ppm-dap.md:1733]: We actually use both! I prefer `truncate` so I switched to that.

File: draft-ietf-ppm-dap.md
Comment (branlwyd) [reply draft-ietf-ppm-dap.md:2433]: aagh, sorry, I keep making that mistake

File: draft-ietf-ppm-dap.md
Comment (branlwyd) [reply draft-ietf-ppm-dap.md:2437]: 👍🏻 It turned out there were a bunch of places that got this wrong -- I fixed them all up. There were a few places where it was ambiguous; for the most part I made as few changes as possible while still referring to "sub-protocols" when discussing only the upload/aggregation/collection portions of DAP.

File: draft-ietf-ppm-dap.md
Comment (branlwyd) [reply draft-ietf-ppm-dap.md:2513]: Hmm, I think the old text was slightly inaccurate for the time-interval case: aggregators don't have to store an aggregate share for each possible a batch interval (i.e., sequence of time windows) -- an aggregator needs to store an aggregate share for each _time window_.

I had felt that distinguishing between the batch modes was not so useful here; I restored the text, modified for accuracy.

File: draft-ietf-ppm-dap.md
Comment (branlwyd) [reply draft-ietf-ppm-dap.md:2477]: OK -- I dropped this because we'll now be storing perhaps a few thousand ID ranges per time window, but no objection to keeping this text around.

File: draft-ietf-ppm-dap.md
Comment (branlwyd) [reply draft-ietf-ppm-dap.md:1316]: I think a `MUST` is appropriate here for the Leader: the Leader must follow this suggestion to avoid having otherwise-acceptable reports rejected as a replay.

The Helper can indeed make a few different choices about how to deal with an aggregation job which does intersect with a previous aggregation job. The relevant text for the Helper is currently:

> Check if the report has been previously aggregated. If so, the input share MUST be marked as invalid with the error `report_replayed`. A report is considered previously aggregated if it intersects with a report ID range for a completed aggregation job (see {{report-id-ranges}}).

... which implies to me that we have to effectively allow "merging" of ranges if they intersect, in the case that an intersecting aggregation job is received. (A Helper which does something more specific, e.g. marking individual report IDs as used, would work to avoid replays; but I'm not sure a Leader could take advantage of this without knowing the specific strategy used by the Helper.)

I'm personally happy with this level of flexibility, but maybe the Helper should be more be more flexible? If so, we might consider relaxing the Leader's MUST here.

Comment (branlwyd) [reply draft-ietf-ppm-dap.md:1316]: > I agree that calling the new thing a "report ID range" makes everything hard to read. This sentence looks like it is comparing report ID ranges but it is actually comparing "report ID ranges".

Indeed; I am careful to use the term "interval" where an "interval" is truly needed. But "report ID range" does not capture that a "report ID range" includes a time window, too.

> Considering the fact that you're saying "a set of" here. Does that mean the old behavior would still be allowed by making each range have exactly one report ID?

Not quite -- an aggregation job may include reports in multiple time windows. In this case, there would be one "report ID range" for each time window represented in the aggregation job.

> To clarify, I am wondering why anybody would want to do this given that the report IDs are randomly assigned. Is the intent to allow people to recover from some amount of late arriving reports?

Nope, I think late-arriving reports are effectively lost (with high probability) in this scheme. The "report ID ranges" in use are all "maximal" in the sense that they stretch from the smallest to the largest available report ID.

Comment (cjpatton) [reply draft-ietf-ppm-dap.md:1321]: I prefer MUST here as well.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [reply draft-ietf-ppm-dap.md:2477]: I'm just suggesting we update the text.

File: draft-ietf-ppm-dap.md
Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1321]: I think if the "set" only covers the "different time intervals" part you need to clarify. The way it's written the ranges don't need to be maximal. Somebody could send 3 ranges for t1 and 7 ranges for t2 as longs as they in total cover all the reports. And in the limit, send one range per report and everything turns into linear search again.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:408]: I might also mention briefly what this is used for (replay protection for one!)

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1728]:
```suggestion
Replay protection is implemented using "report ID ranges". A report ID range is an
```

**Review by cjpatton — APPROVED** [review]

**Comment by branlwyd** [comment]
I renamed "report ID range" to "report ID commitment", in an attempt to alleviate confusion around the fact that a "report ID range" contained an "interval of report IDs" (but also a timestamp). Along with this change, a report ID commitment "conflicts" (rather than "intersects") with another report ID commitment.

File: draft-ietf-ppm-dap.md
Comment (branlwyd) [reply draft-ietf-ppm-dap.md]: I think the ranges do need to be maximal; there actually isn't much wiggle room here. 

The relevant Leader text is:

> The set of candidate reports MUST produce a set of report ID commitments which
does not conflict with the report ID commitments for any other aggregation job
associated with the task (see {{report-id-commitments}}).

[note that "report ID range" got renamed to "report ID commitment" here]

The relevant Helper text is:

> 1. Check if the report has been previously aggregated. If so, the input share
   MUST be marked as invalid with the error `report_replayed`. A report is
   considered previously aggregated if it conflicts with a report ID commitment
   for a completed aggregation job (see {{report-id-commitments}}).

That is, both fall back to the report-id-commitments section; this section specifies a maximal range:

> The report ID commitments associated with a given set of reports are determined by the following procedure:
> 
> 1. Partition the set of reports into groups based on the time window their
>    client timestamp falls into; this might be achieved by partitioning the
>    reports on a key generated by truncating each report's timestamp down to the
>    nearest multiple of the task's `time_precision`.
> 2. For each partitioned group of reports, generate an interval from the smallest
>    report ID in the group to the largest report ID in the group, inclusive. The
>    report ID commitment for this partition is this interval, associated with the
>    time window that generated this partition.

I had previously believed there was some wiggle room for what the Helper commits if the Leader fails in its `MUST` & sends an aggregation job containing a report which falls into an existing report ID commitment. However, now I think the Helper's behavior is fully specified here with no wiggle room:

* Any reports conflicting with an existing report ID commitment must be marked as `report_replayed`.
* The new commitment associated with the aggregation job is still the maximal report ID commitments suggested by the report-id-commitments section -- this commitment will conflict with an already-existing commitment, but that's OK. (A clever implementation might merge conflicting commitments without changing its externally-observable behavior, but this isn't required.)

Comment (branlwyd) [reply draft-ietf-ppm-dap.md]: By the way -- I'm interested in your thoughts on the new text in the report-id-commitments section @simon-friedberger. Other than renaming "report ID ranges" to "report ID commitments", I tried to make it clearer that the set of report ID commitments is one-per-time-window, rather than allowing the aggregators to generate a larger count of smaller commitment intervals.

**Comment by mendess** [comment]
I think this change is very positive to the protocol, provided the leader is okay with the latency cost it will pay when using leader-selected. But I think one final tweak should be made

### Problem

Lets assume the storage solution for replay protection is sharding on time the time window of reports. Meaning storage transactions only contend when they are trying to check for replays of two time windows.

If two Aggregation Jobs have reports belonging to the same time window then they will contend on the storage shard used. This gets even worse if the aggregation jobs have a "mixed bag" of time windows, which increases the chance that two jobs might contend on the time window.

### Suggested change

I think the protocol should encourage leaders (non normatively) to avoid having multiple aggregation jobs "share time windows". Put another way: It might create higher resource contention when N aggregation jobs contain reports belonging to the same time window.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:625]: Orthongonal nit: back when we got a review from HTTPDIR, we got a recommendation to use the term "interaction" or "sub-interaction" instead of "sub-protocol". Mostly this was cleaned up in #450, but I appear to have missed this one.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:923]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1061]: I don't object to using the verb "truncate" here (is the precedent [Go's `time.Truncate`](https://pkg.go.dev/time#Time.Truncate)?), but "truncate [...] down" is awkward. There's no such thing as "truncating up" so why say "truncate down"?
```suggestion
      truncate this value to the nearest multiple of the task's
```

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1162]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1166]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1175]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1248]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1730]: An interval of time makes sense, but an interval of report IDs is a strange notion.
```suggestion
commitment is an ordered set of report IDs (using lexicographic ordering),
```

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1742]:
```suggestion
   reports on a key generated by truncating each report's timestamp to the
```

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1943]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:2424]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:2431]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:2441]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:2769]: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

**Review by tgeoghegan — COMMENTED** [review]
This is an editorial review, and I will defer to those who are knee-deep in implementation on whether the change is a good idea.

**Comment by cjpatton** [comment]
Closed in favor of #564. The Leader must wait until the current time window has elapsed before it can begin aggregating. This means the latency of collecting the aggregate result depends on the `time_precision` parameter for the task, which maybe something like an hour. See https://mailarchive.ietf.org/arch/msg/ppm/mPYFQ9-AjfbSZ5r67gJyJ4cSaBw/ for details.

**Pull request closed without merging**
